// vue-promise-dialogs/dist/index.mjs
import { ref, set, markRaw, del, defineComponent, onBeforeUnmount, h, isVue3 } from "vue-demi";
const wrapperExists = /* @__PURE__ */ ref(false);
const dialogsData = /* @__PURE__ */ ref({});
function add(component, params, unmountDelay) {
  return new Promise((resolve, reject) => {
    set(dialogsData.value, Symbol(), {
      component: markRaw(component),
      params,
      promiseResolve: resolve,
      promiseReject: reject,
      unmountDelay
    });
  });
}
function resolveDialog(id, result, unmountDelay) {
  dialogsData.value[id].promiseResolve(result);
  unmountDialog(id, unmountDelay);
}
function rejectDialog(id, error, unmountDelay) {
  dialogsData.value[id].promiseReject(error);
  unmountDialog(id, unmountDelay);
}
function unmountDialog(id, delay) {
  const unmount = () => del(dialogsData.value, id);
  if (delay) {
    setTimeout(unmount, delay);
  } else {
    unmount();
  }
}
var PromiseDialogsWrapper = /* @__PURE__ */ defineComponent({
  name: "PromiseDialogsWrapper",
  props: {
    unmountDelay: {
      type: Number,
      default: 0
    }
  },
  setup(props) {
    if (wrapperExists.value) {
      console.error("PromiseDialogsWrapper instance already exists");
    }
    wrapperExists.value = true;
    onBeforeUnmount(() => {
      wrapperExists.value = false;
      dialogsData.value = {};
    });
    return () => h("div", Object.getOwnPropertySymbols(dialogsData.value).map((id) => {
      const value = dialogsData.value[id];
      const component = value.component;
      const params = value.params;
      const resolve = (result, unmountDelay) => resolveDialog(id, result, unmountDelay || value.unmountDelay || props.unmountDelay);
      const reject = (error, unmountDelay) => rejectDialog(id, error, unmountDelay || value.unmountDelay || props.unmountDelay);
      if (isVue3) {
        return h(component, {
          key: id,
          params,
          onResolve: resolve,
          onReject: reject
        });
      }
      return h(component, {
        key: id,
        props: {
          params
        },
        on: {
          resolve,
          reject
        }
      });
    }));
  }
});
function createPromiseDialog(component, unmountDelay) {
  return function dialogFunction(params) {
    if (!wrapperExists.value) {
      throw new Error("PromiseDialogsWrapper instance not found");
    }
    return add(component, params, unmountDelay);
  };
}
function closeAllDialogs(reason, unmountDelay) {
  Object.getOwnPropertySymbols(dialogsData.value).forEach((id) => rejectDialog(id, reason, unmountDelay));
}
export { PromiseDialogsWrapper, closeAllDialogs, createPromiseDialog };
